File: CPP/src/BasicLBP.cpp
#include "LBPBase.hpp"

namespace LBP
{

namespace detail
{


BasicLBP::BasicLBP()
{

}
BasicLBP::~BasicLBP()
{

}
bool BasicLBP::run_mutable(const cv::Mat& src, cv::Mat& dst)
{
	int radius = 1;
	int rows = src.rows;
	int cols = src.cols;
	if (rows < 2 * radius || cols < 2 * radius || src.type() != 0) // 只支持灰度图
		return false;

	dst = cv::Mat(rows - 2 * radius, cols - 2 * radius, CV_8UC1);
	int x(0), y(0);
	int neighbor = 8;
	typedef unsigned char T;
	T center(0), edges(0);
	int pos_x[] = { -1, -1, 0, 1, 1, 1, 0, -1 };
	int pos_y[] = { 0, 1, 1, 1, 0, -1, -1, -1 };
	for (int k = 0; k < neighbor; ++k)
	{
		x = pos_x[k];
		y = pos_y[k];
		for (int i = radius; i < src.rows - radius; ++i)
		{
			for (int j = radius; j < src.cols - radius; ++j)
			{
				center = src.at<T>(i, j);
				edges = src.at<T>(i + x, j + y);
				T signal = (edges > center) ? 1 : 0;
				dst.at<T>(i - radius, j - radius) |= signal << (neighbor - k - 1);
			}
		}
	}
	return true;
}




}
}

File: RUST/src/BasicLBP.rs
use super::LBPBase::IBase;
extern crate image;
use image::{GrayImage, Luma, Pixel};

pub struct BasicLBP
{

}

impl IBase for BasicLBP
{
    fn run_mutable(&self, src : &GrayImage) -> GrayImage
    {
        // let dst : GrayImage = GrayImage::new(1,1);
        let radius : u32 = 1;
        let offset : u32 = 0;
        let (rows, cols) : (u32, u32) = src.dimensions();
        let (t_rows, t_cols) : (u32, u32) = (rows - 2 * radius, cols - 2 * radius);

        let mut dst = GrayImage::new(t_rows, t_cols);

        let neighbor : u32 = 8;
        let row_vec : [i32; 8] = [-1, -1, 0, 1, 1, 1, 0, -1];
        let col_vec : [i32; 8] = [0, 1, 1, 1, 0, -1, -1, -1];

        for k in 0 .. neighbor
        {
            if k == 0
            {
                for i in 0 .. t_rows
                {
                    for j in 0 .. t_cols
                    {
                        dst.get_pixel_mut(i, j).channels_mut()[0] = 0u8;
                    }
                }
            }

            for i in 0 .. t_rows
            {
                for j in 0 .. t_cols
                {
                    let cmp_x : u32 = (i as i32 + radius as i32 + row_vec[k as usize]) as u32;
                    let cmp_y : u32 = (j as i32 + radius as i32 + col_vec[k as usize]) as u32;
                    let src_cmp = src.get_pixel(cmp_x, cmp_y);
                    let src_cur = src.get_pixel(i + radius, j + radius);
                    
                    let dst_p = dst.get_pixel_mut(i, j).channels_mut();
                    if src_cur[0] > src_cmp[0]
                    {
                        dst_p[0] = (dst_p[0] << 1) ^ 1u8;
                    }
                    else 
                    {
                        dst_p[0] = (dst_p[0] << 1) ^ 0u8;
                    }
                }
            }


        }
        return dst;
    }
}